### PR TITLE
CLN: cleanup `Model.n_inputs` and `Model.n_outputs` properties

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
This is a follow up to #9298, and cleans up code that was introduced in 4.0 to mitigate a deprecation that was concluded in 5.0

For context, I discovered this while working on #16630, from which this is extracted.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
My approach isn't quite right: removing these properties altogether makes the instance attribute writable when it clearly shouldn't be.

**Comment by pllim** [comment]
@nden or @WilliamJamieson , any advice? Thanks!

**Comment by neutrinoceros** [comment]
I should note that I revised this PR since my previous comment. This is now ready for review.

File: astropy/modeling/core.py
Comment (pllim) [inline astropy/modeling/core.py:806]: Should we not keep `return self.__class__.n_inputs`?

Comment (neutrinoceros) [reply astropy/modeling/core.py:806]: I know that's what the comment said but empirically: no. That branch was actually never reached in any test.

**Comment by pllim** [comment]
I feel like this needs a change log because some "deprecated behavior" is now removed but let's wait for `modeling` maintainers to chime in. Thanks!

**Comment by neutrinoceros** [comment]
Supposedly it's just a clean up and shouldn't be visible to users, but I'll admit this was pretty confusing for me so maybe I misunderstood the implications, in which case I'll happily add a changelog entry !

File: astropy/modeling/tests/test_core.py
Comment (astrofrog) [inline astropy/modeling/tests/test_core.py:374]: Why change this?

Comment (neutrinoceros) [reply astropy/modeling/tests/test_core.py:374]: The comment I'm trying to honor from the `n_outputs` property definition indicates that `output` shouldn't be supported any more as a class attribute. Only a couple test cases were still relying on it, so I updated them for consistency. In effect however, they should still work after my cleanup, given that class attributes can be used as defaults for missing instance attributes.

**Review by WilliamJamieson — APPROVED** [review]
These changes look good to me.

**Comment by pllim** [comment]
@WilliamJamieson , thanks for the review. Does it need a change log?

**Comment by WilliamJamieson** [comment]
It's probably worth adding a change log since it is removing things.

**Comment by neutrinoceros** [comment]
Is it ? from a user perspective I don't know that anything is going away, or what the changelog should say.

**Comment by Cadair** [comment]
I think there should be a changelog for people who write their own models, they don't need to define these properties any more?

**Comment by neutrinoceros** [comment]
My understanding is they already didn't.
It seems this patch is more confusing than I anticipated so let me try to explain what I did a little better:
The base `Model` class currently defines `n_inputs` and `n_outputs` as class attributes *and then* rebinds these names to properties. This means that the class attributes are shadowed immediately and thus never observable for end users. This is why ruff 0.5 is flagging them ([rule F811](https://docs.astral.sh/ruff/rules/redefined-while-unused/)), so removing them doesn't affect users.
While I was in there, I saw these `TODO` comments indicating that complexity in these properties was meant to be eliminated so I followed the bread crumbs and removed everything that wasn't covered by tests, but again, nothing I did there is visible to users.

**Comment by nstarman** [comment]
Are we good to merge this so #16630 can go in as well?

File: astropy/modeling/tests/test_core.py
Comment (nden) [reply astropy/modeling/tests/test_core.py:374]: `outputs`, as well as `inputs`, are user facing attributes which can be changed on instances to reflect better the specific use case by assigning meaningful names. For example, if the input to a gaussian is wavelength, it will be much better for users to see this name instead of `x`.

**Comment by nden** [comment]
I need to review this and run tests. By design the class attributes `n_inputs` and `n_outputs` are the only way to set the number of inputs and outputs, as `inputs` and `outputs` can be changed. However, there are some models (Mapping, Identity) for which the number of in(out)puts can be determined only at the instance level. IIRC, a work around this was to introduce the properties.

Setting `n_inputs` and `n_outputs` in new models is required. I'm hesitant to approve this without a careful review and as a minimum it will require documentation changes.

File: astropy/modeling/tests/test_core.py
Comment (neutrinoceros) [reply astropy/modeling/tests/test_core.py:374]: @nden sorry I might be confused by constant context switching but I can't seem to connect the dots here; are you completing my explanation or are you suggesting a change that I'm missing ?

Comment (nden) [reply astropy/modeling/tests/test_core.py:374]: @neutrinoceros Sorry, I wasn't clear. I meant this needs to be in the public interface as end users can change it.

**Comment by neutrinoceros** [comment]
> Are we good to merge this so https://github.com/astropy/astropy/pull/16630 can go in as well?

@nstarman I've added noqa comments in #16630 and linked to *this* PR so #16630 is not blocked by this one anymore.

@nden there's no rush on my side. I see that the likelihood that I'm the one who's most confused by this change is skyrocketing, but it's still not clear to me that this has any user-visible effects; any visible API change here is unintentional and suffices to justify closing this PR without merge.

**Comment by neutrinoceros** [comment]
rebased to resolve a merge conflict

**Comment by pllim** [comment]
What is the status of this PR? Anything left to resolve?

**Comment by neutrinoceros** [comment]
we never converged on wether this needs a changelog or not.
@Cadair argued there should be one (https://github.com/astropy/astropy/pull/16633#issuecomment-2213618651)
but it's still not clear to me that this has *any* user visible implications (https://github.com/astropy/astropy/pull/16633#issuecomment-2197563387). In any case, this is not *meant* to have any externally observable impact, so as much as I'd fancy writing a changelog just to expedite the PR, I can't do so without a clear understanding of what needs to be told. If anyone insists that there's a user visible change here, it'd help a lot if I could see an example of broken code.

**Comment by pllim** [comment]
@nden @perrygreenfield @WilliamJamieson -- any opinions?

**Comment by neutrinoceros** [comment]
On the off chance that this is actually breaking to anyone (but I still do not think it is), how would we feel about including this in the first release candidate for 7.0 ? if anyone reports anything we should just revert it then.

**Comment by pllim** [comment]
Can u pls rebase again just to make sure CI still green? Wheel jobs got stuck. Thanks!

**Comment by neutrinoceros** [comment]
done !

**Comment by pllim** [comment]
Thanks, all!

**Pull request merged**
